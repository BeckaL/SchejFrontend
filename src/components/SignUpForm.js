import React, { Component } from 'react'
import axios from 'axios'
import { setAuthStorage } from '../actions/authentication'

export default class SignUpForm extends Component {
  constructor(props) {
    super(props)

    this.state = {
      name: "",
      organisation: "",
      email: "",
      mobile: "",
      password: "",
      passwordConfirmation: ""
    }
  }

  handleChange = event => {
    this.setState({
      [event.target.name]: event.target.value
    })
  }

  handleSubmit = event => {
    event.preventDefault()
    if (this.hasNonMatchingPasswords()) {
      alert('Passwords do not match')
    } else if (this.hasInvalidPassword()) {
      alert('Password must be at least 6 characters')
    } else {
      this.createUserRequest()
    }
  }

  createUserRequest = () => {
    const self = this
    axios.post(
      'http://localhost:3001/api/v1/sign_up',
      {
        name: this.state.name,
        organisation: this.state.organisation,
        email: this.state.email,
        mobile: this.state.mobile,
        password: this.state.password,
        passwordConfirmation: this.state.passwordConfirmation
      }
    )
    .then(response => {
      setAuthStorage(response.data)
      self.props.history.push('/shifts')
      console.log(response)
    })
    .catch(error => {
      if (error.response.status === 422) {
        alert('Email already taken')
      } else {
        console.log(error)
      }
    })
  }

  isValidated = () => {
    return this.state.name.length > 0 && this.state.organisation.length > 0 &&
      this.state.email.length > 0 && this.state.mobile.length > 0 &&
      this.state.password.length > 0 && this.state.passwordConfirmation.length > 0
  }

  hasNonMatchingPasswords = () => {
    return this.state.password !== this.state.passwordConfirmation
  }

  hasInvalidPassword = () => {
    return this.state.password.length < 6
  }

  render() {
    return (
      <form className='sign-up-form' onSubmit={this.handleSubmit}>
        <input id='sign-up-name-entry' className='sign-up-entry' type='text' name='name' placeholder='Name' onChange={this.handleChange}/><br />
        <input id='sign-up-organisation-entry' className='sign-up-entry' type='text' name='organisation' placeholder='organisation' onChange={this.handleChange} /><br />
        <input id='sign-up-email-entry' className='sign-up-entry' type='text' name='email' placeholder='Email' onChange={this.handleChange} /><br />
        <input id='sign-up-mobile-entry' className='sign-up-entry' type='text' name='mobile' placeholder='Mobile' onChange={this.handleChange} /><br />
        <input id='sign-up-password-entry' className='sign-up-entry' type='password' name='password' placeholder='Password' onChange={this.handleChange} /><br />
        <input id='sign-up-password-confirmation' className='sign-up-entry' type='password' name='passwordConfirmation' placeholder='Confirm Password' onChange={this.handleChange} /><br /><br />
        <button id='sign-up-submit' className='custom-button' disabled={!this.isValidated()}>Submit</button>
      </form>
    )
  }
}
