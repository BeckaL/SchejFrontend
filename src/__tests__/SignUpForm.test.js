import React from 'react'
import { mount, shallow } from 'enzyme'
import SignUpForm from '../components/SignUpForm'
import axios from "axios"

jest.mock('axios')

describe("sign up form", () => {
  var wrapper

  beforeAll(() => {
    wrapper = shallow(<SignUpForm history = {[]}/>)
  })

  beforeEach(() => {
    localStorage.clear()
  })

  test("it matches the snapshot", () => {
    expect(wrapper).toMatchSnapshot()
  })

  test("it posts to the API of the new user when form is valid", () => {
    const response = {
      data: {
        name: "TestName",
        id: "1",
        organisation: "TestOrganisation",
        job_title: "TestJob"
      }
    }

    axios.post.mockResolvedValue(response)

    const form = wrapper.find('.sign-up-form')
    const fakeEvent = { preventDefault() {} }

    wrapper.instance().hasNonMatchingPasswords = jest.fn()
    wrapper.instance().hasInvalidPassword = jest.fn()

    form.props().onSubmit(fakeEvent)

    expect(axios.post).toHaveBeenCalledTimes(1)
    expect(axios.post).toHaveBeenCalledWith(
      'http://localhost:3001/api/v1/sign_up', {
        name: "",
        email: "",
        organisation: "",
        job_title: "",
        mobile: "",
        password: "",
        passwordConfirmation: ""
      }
    )
  })

  test('It has a disabled submit button by default', () => {
    const button = wrapper.find('#sign-up-submit')

    expect(button.props().disabled).toBeTruthy()
  })

  test('it renders a button to go to the sign up page', () => {
    const mountedWrapper = mount(<SignUpForm history={[]} />)
    const button = mountedWrapper.find('#sign-in-button')

    button.simulate('click')

    expect(mountedWrapper.props().history).toEqual(['/'])
  })

  test('alerts the user when email already taken', () => {
    const result = Promise.reject({ response: { status: 422 } })
    axios.post.mockImplementation(() => result)

    global.alert = jest.fn()

    const form = wrapper.find('.sign-up-form')
    const fakeEvent = { preventDefault() {} }

    wrapper.instance().hasNonMatchingPasswords = jest.fn()
    wrapper.instance().hasInvalidPassword = jest.fn()

    form.props().onSubmit(fakeEvent)

    setTimeout(() => {
      expect(global.alert).toHaveBeenCalledTimes(1)
      expect(global.alert).toHaveBeenCalledWith('Email already taken')
    }, 10)
  })

  test('alerts the user if passwords do not match', () => {
    global.alert = jest.fn()

    const fakeEvent = { preventDefault() {} }
    const form = wrapper.find('.sign-up-form')

    wrapper.instance().hasNonMatchingPasswords.mockImplementation(() => true)

    form.props().onSubmit(fakeEvent)

    expect(global.alert).toHaveBeenCalledWith('Passwords do not match')
  })

  test('alerts the user if password is invalid length', () => {
    global.alert = jest.fn()

    const fakeEvent = { preventDefault() {} }
    const form = wrapper.find('.sign-up-form')

    wrapper.instance().hasNonMatchingPasswords.mockImplementation(() => false)
    wrapper.instance().hasInvalidPassword.mockImplementation(() => true)

    form.props().onSubmit(fakeEvent)

    expect(global.alert).toHaveBeenCalledWith('Password must be at least 6 characters')
  })

  test('It redirects to the shifts calendar if already signed in', () => {
    localStorage['authenticationToken'] = 'TestToken'
    const secondWrapper = mount(<SignUpForm history={[]} />)

    expect(secondWrapper.props().history).toEqual(['/shifts'])
  })

  afterAll(() => {
    localStorage.clear()
    wrapper.unmount()
  })
})
