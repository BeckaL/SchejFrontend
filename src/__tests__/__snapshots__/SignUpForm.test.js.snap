// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sign up form it matches the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SignUpForm
    history={Array []}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <IndexTitle />,
        <form
          className="sign-up-form"
          onSubmit={[Function]}
        >
          <input
            className="sign-up-entry"
            id="sign-up-name-entry"
            name="name"
            onChange={[Function]}
            placeholder="Name"
            type="text"
          />
          <br />
          <input
            className="sign-up-entry"
            id="sign-up-organisation-entry"
            name="organisation"
            onChange={[Function]}
            placeholder="Organisation"
            type="text"
          />
          <br />
          <input
            className="sign-up-entry"
            id="sign-up-job-title-entry"
            name="jobTitle"
            onChange={[Function]}
            placeholder="Job Title"
            type="text"
          />
          <br />
          <input
            className="sign-up-entry"
            id="sign-up-email-entry"
            name="email"
            onChange={[Function]}
            placeholder="Email"
            type="text"
          />
          <br />
          <input
            className="sign-up-entry"
            id="sign-up-mobile-entry"
            name="mobile"
            onChange={[Function]}
            placeholder="Mobile"
            type="text"
          />
          <br />
          <input
            className="sign-up-entry"
            id="sign-up-password-entry"
            name="password"
            onChange={[Function]}
            placeholder="Password"
            type="password"
          />
          <br />
          <input
            className="sign-up-entry"
            id="sign-up-password-confirmation"
            name="passwordConfirmation"
            onChange={[Function]}
            placeholder="Confirm Password"
            type="password"
          />
          <br />
          <br />
          <button
            className="custom-button"
            disabled={true}
            id="sign-up-submit"
          >
            Submit
          </button>
        </form>,
        <button
          className="custom-button"
          id="sign-in-button"
          onClick={[Function]}
        >
          Sign In
        </button>,
      ],
      "id": "sign-up-page",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <input
              className="sign-up-entry"
              id="sign-up-name-entry"
              name="name"
              onChange={[Function]}
              placeholder="Name"
              type="text"
            />,
            <br />,
            <input
              className="sign-up-entry"
              id="sign-up-organisation-entry"
              name="organisation"
              onChange={[Function]}
              placeholder="Organisation"
              type="text"
            />,
            <br />,
            <input
              className="sign-up-entry"
              id="sign-up-job-title-entry"
              name="jobTitle"
              onChange={[Function]}
              placeholder="Job Title"
              type="text"
            />,
            <br />,
            <input
              className="sign-up-entry"
              id="sign-up-email-entry"
              name="email"
              onChange={[Function]}
              placeholder="Email"
              type="text"
            />,
            <br />,
            <input
              className="sign-up-entry"
              id="sign-up-mobile-entry"
              name="mobile"
              onChange={[Function]}
              placeholder="Mobile"
              type="text"
            />,
            <br />,
            <input
              className="sign-up-entry"
              id="sign-up-password-entry"
              name="password"
              onChange={[Function]}
              placeholder="Password"
              type="password"
            />,
            <br />,
            <input
              className="sign-up-entry"
              id="sign-up-password-confirmation"
              name="passwordConfirmation"
              onChange={[Function]}
              placeholder="Confirm Password"
              type="password"
            />,
            <br />,
            <br />,
            <button
              className="custom-button"
              disabled={true}
              id="sign-up-submit"
            >
              Submit
            </button>,
          ],
          "className": "sign-up-form",
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "sign-up-entry",
              "id": "sign-up-name-entry",
              "name": "name",
              "onChange": [Function],
              "placeholder": "Name",
              "type": "text",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": "br",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "sign-up-entry",
              "id": "sign-up-organisation-entry",
              "name": "organisation",
              "onChange": [Function],
              "placeholder": "Organisation",
              "type": "text",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": "br",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "sign-up-entry",
              "id": "sign-up-job-title-entry",
              "name": "jobTitle",
              "onChange": [Function],
              "placeholder": "Job Title",
              "type": "text",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": "br",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "sign-up-entry",
              "id": "sign-up-email-entry",
              "name": "email",
              "onChange": [Function],
              "placeholder": "Email",
              "type": "text",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": "br",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "sign-up-entry",
              "id": "sign-up-mobile-entry",
              "name": "mobile",
              "onChange": [Function],
              "placeholder": "Mobile",
              "type": "text",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": "br",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "sign-up-entry",
              "id": "sign-up-password-entry",
              "name": "password",
              "onChange": [Function],
              "placeholder": "Password",
              "type": "password",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": "br",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "sign-up-entry",
              "id": "sign-up-password-confirmation",
              "name": "passwordConfirmation",
              "onChange": [Function],
              "placeholder": "Confirm Password",
              "type": "password",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": "br",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": "br",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Submit",
              "className": "custom-button",
              "disabled": true,
              "id": "sign-up-submit",
            },
            "ref": null,
            "rendered": "Submit",
            "type": "button",
          },
        ],
        "type": "form",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Sign In",
          "className": "custom-button",
          "id": "sign-in-button",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Sign In",
        "type": "button",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <IndexTitle />,
          <form
            className="sign-up-form"
            onSubmit={[Function]}
          >
            <input
              className="sign-up-entry"
              id="sign-up-name-entry"
              name="name"
              onChange={[Function]}
              placeholder="Name"
              type="text"
            />
            <br />
            <input
              className="sign-up-entry"
              id="sign-up-organisation-entry"
              name="organisation"
              onChange={[Function]}
              placeholder="Organisation"
              type="text"
            />
            <br />
            <input
              className="sign-up-entry"
              id="sign-up-job-title-entry"
              name="jobTitle"
              onChange={[Function]}
              placeholder="Job Title"
              type="text"
            />
            <br />
            <input
              className="sign-up-entry"
              id="sign-up-email-entry"
              name="email"
              onChange={[Function]}
              placeholder="Email"
              type="text"
            />
            <br />
            <input
              className="sign-up-entry"
              id="sign-up-mobile-entry"
              name="mobile"
              onChange={[Function]}
              placeholder="Mobile"
              type="text"
            />
            <br />
            <input
              className="sign-up-entry"
              id="sign-up-password-entry"
              name="password"
              onChange={[Function]}
              placeholder="Password"
              type="password"
            />
            <br />
            <input
              className="sign-up-entry"
              id="sign-up-password-confirmation"
              name="passwordConfirmation"
              onChange={[Function]}
              placeholder="Confirm Password"
              type="password"
            />
            <br />
            <br />
            <button
              className="custom-button"
              disabled={true}
              id="sign-up-submit"
            >
              Submit
            </button>
          </form>,
          <button
            className="custom-button"
            id="sign-in-button"
            onClick={[Function]}
          >
            Sign In
          </button>,
        ],
        "id": "sign-up-page",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <input
                className="sign-up-entry"
                id="sign-up-name-entry"
                name="name"
                onChange={[Function]}
                placeholder="Name"
                type="text"
              />,
              <br />,
              <input
                className="sign-up-entry"
                id="sign-up-organisation-entry"
                name="organisation"
                onChange={[Function]}
                placeholder="Organisation"
                type="text"
              />,
              <br />,
              <input
                className="sign-up-entry"
                id="sign-up-job-title-entry"
                name="jobTitle"
                onChange={[Function]}
                placeholder="Job Title"
                type="text"
              />,
              <br />,
              <input
                className="sign-up-entry"
                id="sign-up-email-entry"
                name="email"
                onChange={[Function]}
                placeholder="Email"
                type="text"
              />,
              <br />,
              <input
                className="sign-up-entry"
                id="sign-up-mobile-entry"
                name="mobile"
                onChange={[Function]}
                placeholder="Mobile"
                type="text"
              />,
              <br />,
              <input
                className="sign-up-entry"
                id="sign-up-password-entry"
                name="password"
                onChange={[Function]}
                placeholder="Password"
                type="password"
              />,
              <br />,
              <input
                className="sign-up-entry"
                id="sign-up-password-confirmation"
                name="passwordConfirmation"
                onChange={[Function]}
                placeholder="Confirm Password"
                type="password"
              />,
              <br />,
              <br />,
              <button
                className="custom-button"
                disabled={true}
                id="sign-up-submit"
              >
                Submit
              </button>,
            ],
            "className": "sign-up-form",
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "sign-up-entry",
                "id": "sign-up-name-entry",
                "name": "name",
                "onChange": [Function],
                "placeholder": "Name",
                "type": "text",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": "br",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "sign-up-entry",
                "id": "sign-up-organisation-entry",
                "name": "organisation",
                "onChange": [Function],
                "placeholder": "Organisation",
                "type": "text",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": "br",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "sign-up-entry",
                "id": "sign-up-job-title-entry",
                "name": "jobTitle",
                "onChange": [Function],
                "placeholder": "Job Title",
                "type": "text",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": "br",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "sign-up-entry",
                "id": "sign-up-email-entry",
                "name": "email",
                "onChange": [Function],
                "placeholder": "Email",
                "type": "text",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": "br",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "sign-up-entry",
                "id": "sign-up-mobile-entry",
                "name": "mobile",
                "onChange": [Function],
                "placeholder": "Mobile",
                "type": "text",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": "br",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "sign-up-entry",
                "id": "sign-up-password-entry",
                "name": "password",
                "onChange": [Function],
                "placeholder": "Password",
                "type": "password",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": "br",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "sign-up-entry",
                "id": "sign-up-password-confirmation",
                "name": "passwordConfirmation",
                "onChange": [Function],
                "placeholder": "Confirm Password",
                "type": "password",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": "br",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": "br",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Submit",
                "className": "custom-button",
                "disabled": true,
                "id": "sign-up-submit",
              },
              "ref": null,
              "rendered": "Submit",
              "type": "button",
            },
          ],
          "type": "form",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Sign In",
            "className": "custom-button",
            "id": "sign-in-button",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Sign In",
          "type": "button",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
